src/swap.rs

```diff
@@ -415,397 +31,112 @@ where
-    }
-}
-
-#[async_trait::async_trait]
-impl<AW, BW, E> SafeToRedeem for WalletAlice<AW, BW, E>
-where
-    AW: Send + Sync,
-    BW: LatestBlock<Block = ethereum::Block>,
-    E: Send + Sync,
-{
-    async fn is_safe_to_redeem(&self, beta_expiry: Timestamp) -> anyhow::Result<bool> {
-        let ethereum_time = ethereum_latest_time(&self.beta_wallet).await?;
-        // TODO: Apply a buffer depending on the blocktime and how
-        // safe we want to be
-
-        Ok(beta_expiry > ethereum_time)
-    }
-}
-
-#[derive(Debug)]
-pub struct BitcoinWallet {
-    connector: Arc<comit::btsieve::bitcoin::BitcoindConnector>,
-}
-
-impl BitcoinWallet {
-    pub async fn fund(
-        &self,
-        _action: hbit::SendToAddress,
-    ) -> anyhow::Result<hbit::CorrectlyFunded> {
-        todo!()
-    }
-
-    pub async fn redeem(
-        &self,
-        _action: hbit::BroadcastSignedTransaction,
-    ) -> anyhow::Result<hbit::Redeemed> {
-        todo!()
-    }
-
-    pub async fn refund(
-        &self,
-        _action: hbit::BroadcastSignedTransaction,
-    ) -> anyhow::Result<hbit::Refunded> {
-        todo!()
-    }
-}
-
+/// Determine whether funding a smart contract is safe.
+///
+/// Implementations should decide based on blockchain time and
+/// Beta expiry, the shorter of the two expiries.
 #[async_trait::async_trait]
-impl LatestBlock for BitcoinWallet {
-    type Block = bitcoin::Block;
-    async fn latest_block(&self) -> anyhow::Result<Self::Block> {
-        self.connector.as_ref().latest_block().await
-    }
-}
-
-#[derive(Debug)]
-pub struct EthereumWallet {
-    connector: Arc<comit::btsieve::ethereum::Web3Connector>,
-}
-
-impl EthereumWallet {
-    pub async fn deploy(
-        &self,
-        _action: herc20::DeployContract,
-    ) -> anyhow::Result<herc20::Deployed> {
-        todo!()
-    }
-
-    pub async fn fund(
-        &self,
-        _action: herc20::CallContract,
-    ) -> anyhow::Result<herc20::CorrectlyFunded> {
-        todo!()
-    }
-
-    pub async fn redeem(&self, _action: herc20::CallContract) -> anyhow::Result<herc20::Redeemed> {
-        todo!()
-    }
-
-    pub async fn refund(&self, _action: herc20::CallContract) -> anyhow::Result<herc20::Refunded> {
-        todo!()
-    }
+pub trait SafeToFund {
+    async fn is_safe_to_fund(&self, beta_expiry: Timestamp) -> anyhow::Result<bool>;
 }
 
+/// Determine whether redeeming a smart contract is safe.
+///
+/// Implementations should decide based on blockchain time and
+/// expiries.
 #[async_trait::async_trait]
-impl LatestBlock for EthereumWallet {
-    type Block = ethereum::Block;
-    async fn latest_block(&self) -> anyhow::Result<Self::Block> {
-        self.connector.latest_block().await
-    }
-}
-
-async fn ethereum_latest_time<C>(connector: &C) -> anyhow::Result<Timestamp>
-where
-    C: LatestBlock<Block = ethereum::Block>,
-{
-    let timestamp = connector.latest_block().await?.timestamp.into();
-
-    Ok(timestamp)
+pub trait SafeToRedeem {
```

Comment:
```suggestion
pub trait IsSafeToRedeem {
```

Reply:
As previously discussed, I don't think `safe` is the correct terminology.
To me, it's `it's not time to refund` would be a better descriptor of the behaviour I expect.